src/main/scala/com/atomist/rug/runtime/js/PlanBuilder.scala

```diff
@@ -0,0 +1,149 @@
+      case _ => None
+    }
+    Presentable(instruction,label)
+  }
+
+  def constructRespondable(jsRespondable: ScriptObjectMirror): Respondable = {
+    val instruction: Instruction = jsRespondable.getMember("instruction") match {
+      case u: Undefined =>
+        throw new IllegalArgumentException(s"No instruction found in $jsRespondable")
+      case o: ScriptObjectMirror =>
+        constructInstruction(o)
+    }
+    val onSuccess: Option[Callback] = constructCallback(jsRespondable.getMember("onSuccess"))
+    val onError: Option[Callback] = constructCallback(jsRespondable.getMember("onError"))
+    Respondable(instruction, onSuccess, onError)
+  }
+
+  def constructInstruction(jsInstruction: ScriptObjectMirror): Instruction = {
+    val jsInstructionKind = jsInstruction.getMember("kind") match {
+      case kind: String => kind
+      case x => throw new InvalidHandlerResultException(s"An Instruction must have a valid 'kind', actually: $x")
+    }
+    val detail = constructInstructionDetail(jsInstruction)
+    Instruction.from(jsInstructionKind, detail)
+  }
+
+  def constructInstructionDetail(jsInstruction: ScriptObjectMirror): Instruction.Detail = {
+      val parameters: Seq[ParameterValue] = jsInstruction.getMember("parameters") match {
+        case u: Undefined => Nil
+        case o: ScriptObjectMirror =>
+          o.keySet().toArray.toList.map { key =>
+            val name = key.asInstanceOf[String]
+            val value = o.getMember(name)
+            SimpleParameterValue(name, value)
+          }
+      }
+      val (name, coordinates) = jsInstruction.getMember("name") match {
+        case name: String =>
+          (name, None)
+        case o: ScriptObjectMirror =>
+          val name = o.getMember("name").asInstanceOf[String]
+          val coordinates = MavenCoordinate(
+            o.getMember("group").asInstanceOf[String],
+            o.getMember("artifact").asInstanceOf[String],
+            o.getMember("version") match {
+              case v: String => Some(v)
+              case _ => None
+            }
+          )
+          (name, Some(coordinates))
+      }
+      Instruction.Detail(
+        name,
+        coordinates,
+        parameters
+      )
+  }
+
+  def constructCallback(callback: Object): Option[Callback] = {
+    callback match {
+      case u: Undefined => None
+      case jsOnSuccess: ScriptObjectMirror =>
+        (jsOnSuccess.hasMember("text"), jsOnSuccess.hasMember("body"), jsOnSuccess.hasMember("kind"), jsOnSuccess.hasMember("messages")) match {
+          case (true, _, _, _) => Some(constructMessage(jsOnSuccess))
+          case (false, true, _, _) => Some(constructMessage(jsOnSuccess))
+          case (false, false, true, _) => Some(Respond(constructInstructionDetail(jsOnSuccess)))
+          case (false, false, false, true) => Some(constructPlan(jsOnSuccess))
+          case _ => throw new InvalidHandlerResultException(s"Cannot create CallBack from: $jsOnSuccess")
```

Comment:
Would an `if`, `else if`, ..., `else` structure be more clear here? It would be more efficient I would think.